### DISCUSSION TITLE
How do I pass around java objects without using them ?

### DISCUSSION THREAD

**Discussion by bananasmoothii** [caoccao/Javet#250] [Q&A]
I have a java class `Player` that I want to store in js variables and pass in intercepted functions but without using properties or functions directly on `Player`. Do I still have to use `JavetProxyConverter` ?

**Comment by caoccao** [comment]
Good question. No, you don't have to. There are many ways of doing so. Here is an example.

1. Declare a static `Map<Integer, Player> map`.
2. Generate an unique `Integer` key per `Player` and put them to the `map`.
3. Pass the key to V8.
4. Get the key from the V8 function and get the `Player` by the key.

**Reply by bananasmoothii** [reply]
Ok thanks. So I can't pass anything else than primitives or maps and lists without `JavetProxyConverter`.

My case is that I have a `@V8Function` returning a `Collections<Player>` (about 50-100 items) that the script directly passes to another function `@V8Function` taking an `Iterable<Player>`. But in that last (java) function, my iterable has the correct length but all values are replaced with `null`

**Reply by caoccao** [reply]
Could you try `List<Player>` at both side?

**Reply by bananasmoothii** [reply]
It didn't work. Here are my functions:
```kotlin
@Suppress("unused")
object MinecraftJSFuns {
    @V8Function
    fun getAllPlayers(): List<Player> = Bukkit.getOnlinePlayers().toList()

    @V8Function
    fun sendMiniMessage(players: List<Player>, message: String) {
        players.forEach {
            it msg message
        }
    }
}
```
registered like this
```kotlin
val v8Runtime: V8Runtime = V8Host.getV8Instance().createV8Runtime<V8Runtime>().apply {
        createV8ValueObject().use {
            globalObject.set("mc", it)
            it.bind(MinecraftJSFuns)
        }
    }
```
(If you don't know kotlin, basically the `createV8ValueObject` is on the returned value of `createV8Runtime`, and here `use` is the same as `try (var it = v8Runtime.createV8ValueObject()) { ... }`, and `MinecraftJSFuns` is directly a class instance. The `=` in `getAllPlayers` is a shortcut for brackets and `return`)

The code is basically just
```js
mc.sendMiniMessage(mc.getAllPlayers(), "hello")
```
But I still get a List with null elements in `sendMiniMessage`.

However, this does get fixed when using `JavetProxyConverter()`.

**Reply by caoccao** [reply]
So, it works, right?

**Reply by bananasmoothii** [reply]
Yes, but not without converting all objects.
I just hope this won't have too much of a performance impact

**Reply by caoccao** [reply]
You are welcome profiling that. The performance impact is subtle.
